prototypes/ScheduledMergesTest.hs

```diff
@@ -176,29 +177,72 @@ test_merge_again_with_incoming =
 -- properties
 --
 
+-- TODO: also generate nested unions?
+
 -- | Supplying enough credits for the remaining debt completes the union merge.
-prop_union :: [[(LSM.Key, LSM.Op)]] -> Property
-prop_union kopss = length (filter (not . null) kopss) > 1 QC.==>
+prop_union_supply_all :: [[(LSM.Key, LSM.Op)]] -> Property
+prop_union_supply_all kopss = length (filter (not . null) kopss) > 1 QC.==>
     QC.ioProperty $ runWithTracer $ \tr ->
       stToIO $ do
         ts <- traverse (mkTable tr) kopss
         t <- LSM.unions ts
 
         debt@(UnionDebt x) <- LSM.remainingUnionDebt t
-        _ <- LSM.supplyUnionCredits t (UnionCredits x)
+        leftovers <- LSM.supplyUnionCredits t (UnionCredits x)
         debt' <- LSM.remainingUnionDebt t
 
         rep <- dumpRepresentation t
         return $ QC.counterexample (show (debt, debt')) $ QC.conjoin
-          [ debt =/= UnionDebt 0
-          , debt' === UnionDebt 0
+          [ QC.counterexample "debt" $ debt =/= UnionDebt 0
+          , QC.counterexample "debt'" $ debt' === UnionDebt 0
+          , QC.counterexample "leftovers" $ leftovers >= 0
           , hasUnionWith isCompleted rep
           ]
   where
     isCompleted = \case
         MLeaf{} -> True
         MNode{} -> False
 
+-- | The union level will get merged into the last regular level once the union
+-- merge is completed and sufficient new entries have been inserted.
+prop_union_merge_into_levels :: [[(LSM.Key, LSM.Op)]] -> Property
+prop_union_merge_into_levels kopss = length (filter (not . null) kopss) > 1 QC.==>
+    QC.forAll arbitrary $ \firstPay ->
+    QC.ioProperty $ runWithTracer $ \tr ->
+      stToIO $ do
+        ts <- traverse (mkTable tr) kopss
+        t <- LSM.unions ts
+
+        -- pay off the union and insert enough that it fits into
+        -- the last level
+        let payOffDebt = do
+              UnionDebt d <- LSM.remainingUnionDebt t
+              _ <- LSM.supplyUnionCredits t (UnionCredits d)
+              return ()
+
+        -- insert as many new entries as there are in the completed
+        -- union level
+        let fillTable = do
+              unionRunSize <- length <$> LSM.logicalValue t
+              LSM.inserts tr t
+                [(K k, V 0, Nothing) | k <- [1 .. unionRunSize]]
```

Comment:
What do we need these inserts for?

Reply:
To make sure the completed union fits into the last level. Otherwise it currently won't be migrated, although that could change with https://github.com/IntersectMBO/lsm-tree/pull/709#discussion_r2077543468.